tests/cases/compiler/nonNullFullInference.ts

```diff
@@ -0,0 +1,17 @@
+// @noImplicitAny: true
+// https://github.com/microsoft/TypeScript/issues/19577
+
+function testNonNullInference(numbers: number[]) {
+    let last;
+
+    for (const n of numbers) {
+        if (n % 2) {
+            return n;
+        }
+
+        last = n;
+    }
+
+    last; // number
+    last!; // number
+}
```

Comment:
I feel like this isn't enough tests. Does anyone have suggestions for tests to add?

Reply:
I'm not sure if this is any different from the current test case - there is a good chance that it isn't - but I've seen that your changes are also related to `autoArrayType` so I've tried to use it in a test case. 

https://www.typescriptlang.org/play?ts=4.7.4#code/GYVwdgxgLglg9mABFApgZygOQZkAbPASTGBQCcVIUAKMEAWwCNy0AuROp8gbQF0BKRAG8AUInGI8KKIgpp8UMRIgIMiAIZkyiALyI+IpeOBxt1FWDVI4wDg2Zk0g0RNeIYt2ogCkiAEzORm6uFFAgZEjcYLwA3EGuAL6GweKaZAB0AA4gaAAWtPzxEnIKuhpaQUlBJXhQca4A9A2IAHoA-NXoCgCEQU2tHZ1hEbJdtSIJQA